internal/transport/http2_client.go

```diff
@@ -252,6 +253,9 @@ func newHTTP2Client(connectCtx, ctx context.Context, addr TargetInfo, opts Conne
 		t.channelzID = channelz.RegisterNormalSocket(t, opts.ChannelzParentID, "")
 	}
 	if t.kp.Time != infinity {
+		if err = syscall.SetTCPUserTimeout(t.conn, kp.Timeout); err != nil {
```

Comment:
This will need to be done before handshaking ~line 170, otherwise you won't have a TCPConn anymore.

Also, you need to only do this for TCP connections.  The dialer could return a net.Conn backed by a UDS instead.